runtime/health.go

```diff
@@ -15,26 +16,17 @@ type Health struct {
 }
 
 type InstanceHealth struct {
-	Controller error
-	Repo       error
-	OLAP       error
-}
+	// always recomputed
+	Controller        string
+	Repo              string
+	ParseErrCount     int
+	ReconcileErrCount int
 
-func (h *InstanceHealth) To() *runtimev1.InstanceHealth {
-	if h == nil {
-		return nil
-	}
-	r := &runtimev1.InstanceHealth{}
-	if h.Controller != nil {
-		r.ControllerError = h.Controller.Error()
-	}
-	if h.Repo != nil {
-		r.RepoError = h.Repo.Error()
-	}
-	if h.OLAP != nil {
-		r.OlapError = h.OLAP.Error()
-	}
-	return r
+	// cached
+	OLAP         string
+	MetricsViews map[string]metricsViewHealth
+
+	ControllerVersion int64
 }
```

Comment:
Currently it keeps it keeps cached fields separated from non cached ones which seems beneficial to quickly check what is cached.
```
// InstanceHealth contains health information for a single instance.
// The information about OLAP and metrics views is cached in the catalog.
// We want to avoid hitting the underlying OLAP engine when OLAP engine can scale to zero when no queries are generated within TTL.
// We do not want to keep it running just to check health. In such cases, we use the cached health information.
type InstanceHealth struct {
	// always recomputed on every health check
	Controller        string `json:"-"`
	Repo              string `json:"-"`
	ParseErrCount     int    `json:"-"`
	ReconcileErrCount int    `json:"-"`

	// cached health check information can be used if controller version is same and metrics view spec is same
	OLAP              string                                    `json:"olap"`
	MetricsViews      map[string]InstanceHealthMetricsViewError `json:"metrics_views"`
	ControllerVersion int64                                     `json:"controller_version"`
}
```

Reply:
1. I agree it's a noteworthy implementation detail, but I'm not sure it's the main thing guiding the API/interface design here. The approach of the proto message where fields are ordered roughly be severity of the error makes more sense to me, and I think would make sense to mirror internally here. The caching behavior could alternatively be noted in per-field comments instead of a section-level comment.

2. Something else as well – I'll propose not hiding any of the fields from the JSON output, even those that are recomputed. It will be useful to have the DB have an up-to-date snapshot of the latest health information across all these properties.